_posts/2015-10-22-transcending-rest-and-rpc.md

```diff
@@ -0,0 +1,102 @@
+
+They based that platform in the idea of JSON Graph. Let me explain it briefly. JSON is in essence a tree structure. However, most of our data has relationships so it should be modeled like a graph. Using REST design programmers usually solve that problem through ids, but that makes our live really difficult if you think in problems like cache invalidation. As they say:
+
+> JSON Graph is valid JSON and can be parsed by any JSON parser. However JSON Graph introduces new primitive types to JSON to allow JSON to be used to represent graph information in a simple and consistent way.
+
+They borrow the idea from unix filesystem. That filesystem is a tree, but thanks to symlinks you can emulate the behaviour of a graph. That has massive implications in the way that Falcor optimises calls to backend services. Instead of keeping some values in a denormalised fashion, Falcor keeps references to that normalised value, so, if you ask for a value that has been already retrieved, Falcor will avoid that call. Also that makes cache invalidation business much more easier.
+
+<img src="/assets/img/custom/blog/falcor-services-diagram.png" />
+
+As you can see in the example, we don't store denormalised documents for titles inside of genres, but a reference through a titles map indexed by id.
+I can show you some code that I've been working on for a proof of concept. Falcor server side is currently only available on Node.js but there are plans to porting it into another platforms.
+
+```
+router.get('/model.json', falcorKoa.dataSourceRoute(new FalcorRouter([
+		{
+			route: "teamsById[{integers:teamIds}]['name', 'memberCount']",
+			get  : function(pathSet) {
+				const teamIds = pathSet.teamIds;
+				const keys = pathSet[2];
+				return teamsService.getTeams()
+						.then(function(teams) {
+							var results = [];
+							teamIds.forEach(function(teamId) {
+								keys.forEach(function(key) {
+									var team = teams[teamId];
+									results.push({
+										path : ['team', teamId, key],
+										value: team[key]
+									});
+								});
+							});
+							return results;
+						});
+			}
+		}
+	])));
+
+```
+
+The client side would look like this:
+
+```
+const model = new falcor.Model({ source: new falcor.HttpDataSource('/model.json') });
+model.get('teamsById['1234', '456'].name');
+model.get('teamsById['789']['name', 'memberCount']');
+```
+
+## IN CONCLUSION
+
+There are not silver bullets in software and nobody is claiming than Falcor or GraphQL are going to solve every single problem with client/server integration business. However, if you have an application with a lot of clients and your backend has a complex data model, it might be worth giving this new paradigm a try.
+
+* [Falcor](http://netflix.github.io/falcor)
+* [GraphQL](https://facebook.github.io/react/blog/2015/05/01/graphql-introduction.html)
+* [Demand-Driven Architecture](http://www.infoq.com/presentations/domain-driven-architecture)
```

Comment:
I personally like to have these links in the text, but this is your personal style so don't change it just on my account :)